File: windows/Marquee.Core.Tests/LastActiveTests.cs
using Marquee.Core.Models;
using Marquee.Core.Tests.Support;

namespace Marquee.Core.Tests;

// HouseholdMember.lastActiveAt (api-v1.md section 11) and the website's
// lastActiveLabel (lib/users/last-active-label.ts) at each threshold.

public sealed class LastActiveTests
{
    private static readonly DateTimeOffset Now = Json.ParseDate("2026-09-25T18:42:10.000Z")!.Value;

    private static string Label(TimeSpan ago) => LastActiveLabel.Format(Now - ago, Now, TimeZoneInfo.Utc);

    [Fact]
    public void FixtureCarriesLastActive()
    {
        var member = Fixtures.Decode<HouseholdMember>("household-member");

        Assert.True(member.ReportsLastActive);
        Assert.Equal(Now, member.LastActiveAt);
        Assert.Equal("Active now", member.LastActiveLine(Now));
    }

    [Fact]
    public void PresentNullIsNeverSignedIn()
    {
        var json = Fixtures.Read("household-member")
            .Replace("\"lastActiveAt\": \"2026-09-25T18:42:10.000Z\"", "\"lastActiveAt\": null", StringComparison.Ordinal);
        Assert.Contains("\"lastActiveAt\": null", json, StringComparison.Ordinal);
        var member = Json.Decode<HouseholdMember>(json);

        Assert.True(member.ReportsLastActive);
        Assert.Null(member.LastActiveAt);
        Assert.Equal("Never signed in", member.LastActiveLine(Now));
    }

    [Fact]
    public void OlderServerShowsNothing()
    {
        var json = Fixtures.Read("household-member")
            .Replace("\"lastActiveAt\": \"2026-09-25T18:42:10.000Z\"", "\"legacy\": 0", StringComparison.Ordinal);
        Assert.DoesNotContain("lastActiveAt", json, StringComparison.Ordinal);
        var member = Json.Decode<HouseholdMember>(json);

        Assert.False(member.ReportsLastActive);
        Assert.Null(member.LastActiveAt);
        Assert.Null(member.LastActiveLine(Now));
    }

    [Fact]
    public void UserListCarriesLastActive()
    {
        var member = Assert.Single(Fixtures.Decode<ListResponse<HouseholdMember>>("users").Results);

        Assert.True(member.ReportsLastActive);
    }

    [Fact]
    public void NeverSignedIn() => Assert.Equal("Never signed in", LastActiveLabel.Format(null, Now));

    [Fact]
    public void ActiveNowUnderTenMinutes()
    {
        Assert.Equal("Active now", Label(TimeSpan.Zero));
        Assert.Equal("Active now", Label(TimeSpan.FromMinutes(10) - TimeSpan.FromSeconds(1)));
        // A clock slightly behind the server's.
        Assert.Equal("Active now", Label(TimeSpan.FromMinutes(-3)));
    }

    [Fact]
    public void MinutesUnderAnHour()
    {
        Assert.Equal("Active 10 minutes ago", Label(TimeSpan.FromMinutes(10)));
        Assert.Equal("Active 25 minutes ago", Label(TimeSpan.FromMinutes(25) + TimeSpan.FromSeconds(40)));
        Assert.Equal("Active 59 minutes ago", Label(TimeSpan.FromHours(1) - TimeSpan.FromSeconds(1)));
    }

    [Fact]
    public void HoursUnderADay()
    {
        Assert.Equal("Active 1 hour ago", Label(TimeSpan.FromHours(1)));
        Assert.Equal("Active 1 hour ago", Label(TimeSpan.FromMinutes(119)));
        Assert.Equal("Active 2 hours ago", Label(TimeSpan.FromHours(2)));
        Assert.Equal("Active 23 hours ago", Label(TimeSpan.FromDays(1) - TimeSpan.FromSeconds(1)));
    }

    [Fact]
    public void YesterdayUnderTwoDays()
    {
        Assert.Equal("Active yesterday", Label(TimeSpan.FromDays(1)));
        Assert.Equal("Active yesterday", Label(TimeSpan.FromDays(2) - TimeSpan.FromSeconds(1)));
    }

    [Fact]
    public void DaysUnderThirty()
    {
        Assert.Equal("Active 2 days ago", Label(TimeSpan.FromDays(2)));
        Assert.Equal("Active 6 days ago", Label(TimeSpan.FromDays(6.5)));
        Assert.Equal("Active 29 days ago", Label(TimeSpan.FromDays(30) - TimeSpan.FromSeconds(1)));
    }

    [Fact]
    public void DateFromThirtyDays()
    {
        Assert.Equal("Last active Aug 26, 2026", Label(TimeSpan.FromDays(30)));
        var july4 = Json.ParseDate("2026-07-04T12:00:00.000Z")!.Value;
        Assert.Equal("Last active Jul 4, 2026", LastActiveLabel.Format(july4, Now, TimeZoneInfo.Utc));
    }

    [Fact]
    public void DateIsInTheGivenZone()
    {
        var lateUtc = Json.ParseDate("2026-07-05T02:00:00.000Z")!.Value;
        var behindUtc = TimeZoneInfo.CreateCustomTimeZone("UTC-4", TimeSpan.FromHours(-4), "UTC-4", "UTC-4");

        Assert.Equal("Last active Jul 5, 2026", LastActiveLabel.Format(lateUtc, Now, TimeZoneInfo.Utc));
        Assert.Equal("Last active Jul 4, 2026", LastActiveLabel.Format(lateUtc, Now, behindUtc));
    }
}

File: mac/MarqueeTests/LastActiveTests.swift
import XCTest
@testable import Marquee

// Household members' "last active" line: decoding `lastActiveAt` (a date,
// null for "never", or missing from an older server) and the label at each
// threshold of the website's lib/users/last-active-label.ts.

final class LastActiveTests: XCTestCase {
    private let now = Date(timeIntervalSince1970: 1_790_000_000)
    private let utc = TimeZone(identifier: "UTC")!

    private func label(ago seconds: TimeInterval) -> String {
        lastActiveLabel(now.addingTimeInterval(-seconds), now: now, timeZone: utc)
    }

    private func decode(_ json: String) throws -> API.HouseholdMember {
        try APIClient.decoder.decode(API.HouseholdMember.self, from: Data(json.utf8))
    }

    private static let memberBase = #""id":"6f1c2a4e-8b1d-4c3e-9f0a-2b7d5e8c1a90","username":"sam","displayName":null,"role":"member","autoApproveMovies":false,"autoApproveTv":false,"createdAt":"2026-09-01T12:00:00.000Z","isCurrentUser":false"#

    func testFixtureDecodesLastActiveAt() throws {
        let url = Bundle(for: Self.self).resourceURL!
            .appendingPathComponent("Fixtures/api/household-member.json")
        let member = try APIClient.decoder.decode(API.HouseholdMember.self, from: Data(contentsOf: url))
        XCTAssertEqual(member.lastActiveAt, ISO8601DateFormatter().date(from: "2026-09-25T18:42:10Z"))
        XCTAssertTrue(member.reportsLastActive)
        XCTAssertEqual(member.lastActiveLine(now: member.lastActiveAt!.addingTimeInterval(3 * 3600)), "Active 3 hours ago")
    }

    func testNullMeansNeverSignedIn() throws {
        let member = try decode("{\(Self.memberBase),\"lastActiveAt\":null}")
        XCTAssertNil(member.lastActiveAt)
        XCTAssertTrue(member.reportsLastActive)
        XCTAssertEqual(member.lastActiveLine(now: now), "Never signed in")
    }

    func testOlderServerShowsNothing() throws {
        let member = try decode("{\(Self.memberBase)}")
        XCTAssertNil(member.lastActiveAt)
        XCTAssertFalse(member.reportsLastActive)
        XCTAssertNil(member.lastActiveLine(now: now))
    }

    func testReencodingKeepsTheField() throws {
        let member = try decode("{\(Self.memberBase),\"lastActiveAt\":\"2026-09-25T18:42:10.000Z\"}")
        let object = try JSONSerialization.jsonObject(with: APIClient.encoder.encode(member)) as? [String: Any]
        XCTAssertNotNil(object?["lastActiveAt"])
        XCTAssertNil(object?["reportsLastActive"])
    }

    func testLabelThresholds() {
        let minute: TimeInterval = 60, hour = 3600.0, day = 86_400.0
        XCTAssertEqual(lastActiveLabel(nil, now: now), "Never signed in")
        XCTAssertEqual(label(ago: 0), "Active now")
        XCTAssertEqual(label(ago: 10 * minute - 1), "Active now")
        XCTAssertEqual(label(ago: 10 * minute), "Active 10 minutes ago")
        XCTAssertEqual(label(ago: 25 * minute + 30), "Active 25 minutes ago")
        XCTAssertEqual(label(ago: hour - 1), "Active 59 minutes ago")
        XCTAssertEqual(label(ago: hour), "Active 1 hour ago")
        XCTAssertEqual(label(ago: 2 * hour - 1), "Active 1 hour ago")
        XCTAssertEqual(label(ago: 2 * hour), "Active 2 hours ago")
        XCTAssertEqual(label(ago: day - 1), "Active 23 hours ago")
        XCTAssertEqual(label(ago: day), "Active yesterday")
        XCTAssertEqual(label(ago: 2 * day - 1), "Active yesterday")
        XCTAssertEqual(label(ago: 2 * day), "Active 2 days ago")
        XCTAssertEqual(label(ago: 30 * day - 1), "Active 29 days ago")
    }

    func testPastThirtyDaysShowsTheDate() {
        let july4 = ISO8601DateFormatter().date(from: "2026-07-04T15:00:00Z")!
        let later = july4.addingTimeInterval(30 * 86_400)
        XCTAssertEqual(lastActiveLabel(july4, now: later, timeZone: utc), "Last active Jul 4, 2026")
        XCTAssertEqual(lastActiveLabel(july4, now: later.addingTimeInterval(365 * 86_400), timeZone: utc), "Last active Jul 4, 2026")
    }
}
